go/kubebuilder/memcached-operator/README.md

```diff
@@ -89,3 +102,4 @@ $ kubectl logs deployment.apps/memcached-operator-controller-manager -n  memcach
 [operator_sdk]: https://github.com/operator-framework/operator-sdk
 [operator_install]: https://sdk.operatorframework.io/docs/install-operator-sdk/
 [quickstart]: https://github.com/operator-framework/operator-sdk/blob/master/website/content/en/docs/kubebuilder/quickstart.md#implement-the-controller
+[certmanager]: https://cert-manager.io/docs/installation/kubernetes/
```

Comment:
I am still unable to deploy: 

```shell
$ make deploy IMG=$IMG
/Users/camilamacedo/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
cd config/manager && kustomize edit set image controller=quay.io/camilamacedo86/memcached-operator:v0.0.1
kustomize build config/default | kubectl apply -f -
namespace/memcached-operator-system unchanged
customresourcedefinition.apiextensions.k8s.io/memcacheds.cache.example.com configured
mutatingwebhookconfiguration.admissionregistration.k8s.io/memcached-operator-mutating-webhook-configuration configured
role.rbac.authorization.k8s.io/memcached-operator-leader-election-role unchanged
clusterrole.rbac.authorization.k8s.io/memcached-operator-manager-role configured
clusterrole.rbac.authorization.k8s.io/memcached-operator-proxy-role unchanged
clusterrole.rbac.authorization.k8s.io/memcached-operator-metrics-reader unchanged
rolebinding.rbac.authorization.k8s.io/memcached-operator-leader-election-rolebinding unchanged
clusterrolebinding.rbac.authorization.k8s.io/memcached-operator-manager-rolebinding unchanged
clusterrolebinding.rbac.authorization.k8s.io/memcached-operator-proxy-rolebinding unchanged
service/memcached-operator-controller-manager-metrics-service unchanged
service/memcached-operator-webhook-service unchanged
deployment.apps/memcached-operator-controller-manager unchanged
certificate.cert-manager.io/memcached-operator-serving-cert unchanged
issuer.cert-manager.io/memcached-operator-selfsigned-issuer unchanged
validatingwebhookconfiguration.admissionregistration.k8s.io/memcached-operator-validating-webhook-configuration configured
kubectl create -f config/samples/cache_v1alpha1_memcached.yaml -n memcached-operator-system 
Error from server (InternalError): error when creating "config/samples/cache_v1alpha1_memcached.yaml": Internal error occurred: failed calling webhook "mmemcached.kb.io": Post https://memcached-operator-webhook-service.memcached-operator-system.svc:443/mutate-cache-example-com-v1alpha1-memcached?timeout=30s: dial tcp 10.98.31.60:443: connect: connection refused
make: [Makefile:47: deploy] Error 1 (ignored)
```

See that I have the cert-manager: 

```
$ kubectl get pods --namespace cert-manager
NAME                                      READY   STATUS    RESTARTS   AGE
cert-manager-7747db9d88-s496m             1/1     Running   0          14m
cert-manager-cainjector-87c85c6ff-t5k4f   1/1     Running   0          14m
cert-manager-webhook-64dc9fff44-gzb2c     1/1     Running   0          14m
```

Reply:
HI @camilamacedo86 I have noticed this error when I have deployed right after work I installed `cert-manager`, Can you run `make deploy` step again in your cluster?

Reply:
Hi @bharathi-tenneti,

I followed the steps to check the cert-manager in the doc:https://cert-manager.io/docs/installation/kubernetes/#verifying-the-installation. also, I indeed tested from a minikube with a new VM just to check that. See: 

```
Status:
  Conditions:
    Last Transition Time:  2020-06-29T16:39:33Z
    Message:               Certificate is up to date and has not expired
    Reason:                Ready
    Status:                True
    Type:                  Ready
  Not After:               2020-09-27T16:39:32Z
Events:
  Type    Reason        Age   From          Message
  ----    ------        ----  ----          -------
  Normal  GeneratedKey  6s    cert-manager  Generated a new private key
  Normal  Requested     6s    cert-manager  Created new CertificateRequest resource "selfsigned-cert-504566127"
  Normal  Issued        5s    cert-manager  Certificate issued successfully
```

Following the full steps: 
```
 $ kubectl apply --validate=false -f https://github.com/jetstack/cert-manager/releases/download/v0.15.1/cert-manager.yaml
customresourcedefinition.apiextensions.k8s.io/issuers.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/orders.acme.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/certificaterequests.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/certificates.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/challenges.acme.cert-manager.io created
customresourcedefinition.apiextensions.k8s.io/clusterissuers.cert-manager.io created
namespace/cert-manager created
serviceaccount/cert-manager-cainjector created
serviceaccount/cert-manager created
serviceaccount/cert-manager-webhook created
clusterrole.rbac.authorization.k8s.io/cert-manager-cainjector created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-orders created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-challenges created
clusterrole.rbac.authorization.k8s.io/cert-manager-edit created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-ingress-shim created
clusterrole.rbac.authorization.k8s.io/cert-manager-view created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-certificates created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-issuers created
clusterrole.rbac.authorization.k8s.io/cert-manager-controller-clusterissuers created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-cainjector created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-orders created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-clusterissuers created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-certificates created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-challenges created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-issuers created
clusterrolebinding.rbac.authorization.k8s.io/cert-manager-controller-ingress-shim created
role.rbac.authorization.k8s.io/cert-manager-cainjector:leaderelection created
role.rbac.authorization.k8s.io/cert-manager:leaderelection created
role.rbac.authorization.k8s.io/cert-manager-webhook:dynamic-serving created
rolebinding.rbac.authorization.k8s.io/cert-manager-cainjector:leaderelection created
rolebinding.rbac.authorization.k8s.io/cert-manager:leaderelection created
rolebinding.rbac.authorization.k8s.io/cert-manager-webhook:dynamic-serving created
service/cert-manager created
service/cert-manager-webhook created
deployment.apps/cert-manager-cainjector created
deployment.apps/cert-manager created
deployment.apps/cert-manager-webhook created
mutatingwebhookconfiguration.admissionregistration.k8s.io/cert-manager-webhook created
validatingwebhookconfiguration.admissionregistration.k8s.io/cert-manager-webhook created
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $   kubectl create clusterrolebinding cluster-admin-binding \
>     --clusterrole=cluster-admin \
>     --user=$(gcloud config get-value core/account)
-bash: gcloud: command not found
clusterrolebinding.rbac.authorization.k8s.io/cluster-admin-binding created
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ kubectl get pods --namespace cert-manager
NAME                                      READY   STATUS              RESTARTS   AGE
cert-manager-7747db9d88-5thnp             0/1     ContainerCreating   0          17s
cert-manager-cainjector-87c85c6ff-2pwqb   1/1     Running             0          17s
cert-manager-webhook-64dc9fff44-sw76t     0/1     Running             0          17s
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ cat <<EOF > test-resources.yaml
> apiVersion: v1
> kind: Namespace
> metadata:
>   name: cert-manager-test
> ---
> apiVersion: cert-manager.io/v1alpha2
> kind: Issuer
> metadata:
>   name: test-selfsigned
>   namespace: cert-manager-test
> spec:
>   selfSigned: {}
> ---
> apiVersion: cert-manager.io/v1alpha2
> kind: Certificate
> metadata:
>   name: selfsigned-cert
>   namespace: cert-manager-test
> spec:
>   dnsNames:
>     - example.com
>   secretName: selfsigned-cert-tls
>   issuerRef:
>     name: test-selfsigned
> EOF
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ cat <<EOF > test-resources.yaml
> apiVersion: v1
> kind: Namespace
> metadata:
>   name: cert-manager-test
> ---
> apiVersion: cert-manager.io/v1alpha2
> kind: Issuer
> metadata:
>   name: test-selfsigned
>   namespace: cert-manager-test
> spec:
>   selfSigned: {}
> ---
> apiVersion: cert-manager.io/v1alpha2
> kind: Certificate
> metadata:
>   name: selfsigned-cert
>   namespace: cert-manager-test
> spec:
>   dnsNames:
>     - example.com
>   secretName: selfsigned-cert-tls
>   issuerRef:
>     name: test-selfsigned
> EOF
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ kubectl apply -f test-resources.yaml
namespace/cert-manager-test created
issuer.cert-manager.io/test-selfsigned created
certificate.cert-manager.io/selfsigned-cert created
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ kubectl describe certificate -n cert-manager-test
Name:         selfsigned-cert
Namespace:    cert-manager-test
Labels:       <none>
Annotations:  API Version:  cert-manager.io/v1alpha3
Kind:         Certificate
Metadata:
  Creation Timestamp:  2020-06-29T16:39:32Z
  Generation:          1
  Managed Fields:
    API Version:  cert-manager.io/v1alpha2
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:dnsNames:
        f:issuerRef:
          .:
          f:name:
        f:secretName:
    Manager:      kubectl
    Operation:    Update
    Time:         2020-06-29T16:39:32Z
    API Version:  cert-manager.io/v1alpha2
    Fields Type:  FieldsV1
    fieldsV1:
      f:status:
        .:
        f:conditions:
        f:notAfter:
    Manager:         controller
    Operation:       Update
    Time:            2020-06-29T16:39:33Z
  Resource Version:  751
  Self Link:         /apis/cert-manager.io/v1alpha3/namespaces/cert-manager-test/certificates/selfsigned-cert
  UID:               bd0803b1-4f4d-44b6-a1bc-6bdae3eb411b
Spec:
  Dns Names:
    example.com
  Issuer Ref:
    Name:       test-selfsigned
  Secret Name:  selfsigned-cert-tls
Status:
  Conditions:
    Last Transition Time:  2020-06-29T16:39:33Z
    Message:               Certificate is up to date and has not expired
    Reason:                Ready
    Status:                True
    Type:                  Ready
  Not After:               2020-09-27T16:39:32Z
Events:
  Type    Reason        Age   From          Message
  ----    ------        ----  ----          -------
  Normal  GeneratedKey  6s    cert-manager  Generated a new private key
  Normal  Requested     6s    cert-manager  Created new CertificateRequest resource "selfsigned-cert-504566127"
  Normal  Issued        5s    cert-manager  Certificate issued successfully
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ make manifests
/Users/camilamacedo/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ make install
/Users/camilamacedo/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
kustomize build config/crd | kubectl apply -f -
customresourcedefinition.apiextensions.k8s.io/memcacheds.cache.example.com created
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ export IMG=quay.io/camilamacedo86/memcached-operator:v0.0.1
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ make docker-build IMG=$IMG
/Users/camilamacedo/go/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
/Users/camilamacedo/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
go test ./... -coverprofile cover.out
?   	github.com/example-inc/memcached-operator	[no test files]
?   	github.com/example-inc/memcached-operator/api/v1alpha1	[no test files]
ok  	github.com/example-inc/memcached-operator/controllers	8.523s	coverage: 0.0% of statements
docker build . -t quay.io/camilamacedo86/memcached-operator:v0.0.1
Sending build context to Docker daemon  138.8kB
Step 1/14 : FROM golang:1.13 as builder
 ---> e025aa1816e5
Step 2/14 : WORKDIR /workspace
 ---> Using cache
 ---> 2dbf60dd90b2
Step 3/14 : COPY go.mod go.mod
 ---> Using cache
 ---> f48fdd6a0100
Step 4/14 : COPY go.sum go.sum
 ---> Using cache
 ---> 1c7cb9d770a2
Step 5/14 : RUN go mod download
 ---> Using cache
 ---> b526edbeb55f
Step 6/14 : COPY main.go main.go
 ---> Using cache
 ---> 78d88297b885
Step 7/14 : COPY api/ api/
 ---> Using cache
 ---> 73ff09442be1
Step 8/14 : COPY controllers/ controllers/
 ---> Using cache
 ---> e9d8a0c89bb5
Step 9/14 : RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 GO111MODULE=on go build -a -o manager main.go
 ---> Using cache
 ---> e98d94687f93
Step 10/14 : FROM gcr.io/distroless/static:nonroot
 ---> 52ed073f70ec
Step 11/14 : WORKDIR /
 ---> Using cache
 ---> 05748550f267
Step 12/14 : COPY --from=builder /workspace/manager .
 ---> Using cache
 ---> 9d98958da419
Step 13/14 : USER nonroot:nonroot
 ---> Using cache
 ---> 2938e16428ff
Step 14/14 : ENTRYPOINT ["/manager"]
 ---> Using cache
 ---> 5594b6a1812d
Successfully built 5594b6a1812d
Successfully tagged quay.io/camilamacedo86/memcached-operator:v0.0.1
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ make docker-push IMG=$IMG
docker push quay.io/camilamacedo86/memcached-operator:v0.0.1
The push refers to repository [quay.io/camilamacedo86/memcached-operator]
cfe9b12364de: Layer already exists 
84ff92691f90: Layer already exists 
42a3027eaac1: Layer already exists 
v0.0.1: digest: sha256:7cc9d157b60c8931f757c04e949b6068f4ca5832f4bbf7dbf471b51801772b63 size: 945
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ make deploy IMG=$IMG
/Users/camilamacedo/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
cd config/manager && kustomize edit set image controller=quay.io/camilamacedo86/memcached-operator:v0.0.1
kustomize build config/default | kubectl apply -f -
namespace/memcached-operator-system created
customresourcedefinition.apiextensions.k8s.io/memcacheds.cache.example.com configured
mutatingwebhookconfiguration.admissionregistration.k8s.io/memcached-operator-mutating-webhook-configuration created
role.rbac.authorization.k8s.io/memcached-operator-leader-election-role created
clusterrole.rbac.authorization.k8s.io/memcached-operator-manager-role created
clusterrole.rbac.authorization.k8s.io/memcached-operator-proxy-role created
clusterrole.rbac.authorization.k8s.io/memcached-operator-metrics-reader created
rolebinding.rbac.authorization.k8s.io/memcached-operator-leader-election-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/memcached-operator-manager-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/memcached-operator-proxy-rolebinding created
service/memcached-operator-controller-manager-metrics-service created
service/memcached-operator-webhook-service created
deployment.apps/memcached-operator-controller-manager created
certificate.cert-manager.io/memcached-operator-serving-cert created
issuer.cert-manager.io/memcached-operator-selfsigned-issuer created
validatingwebhookconfiguration.admissionregistration.k8s.io/memcached-operator-validating-webhook-configuration created
kubectl create -f config/samples/cache_v1alpha1_memcached.yaml -n memcached-operator-system 
Error from server (InternalError): error when creating "config/samples/cache_v1alpha1_memcached.yaml": Internal error occurred: failed calling webhook "mmemcached.kb.io": Post https://memcached-operator-webhook-service.memcached-operator-system.svc:443/mutate-cache-example-com-v1alpha1-memcached?timeout=30s: dial tcp 10.101.161.155:443: connect: connection refused
make: [Makefile:47: deploy] Error 1 (ignored)
camilamacedo@Camilas-MacBook-Pro ~/go/src/github.com/operator-framework/operator-sdk-samples/go/kubebuilder/memcached-operator (upgrade-go) $ 
```

Reply:
yep. Solved.